technic/machines/register/generator.lua

```diff
@@ -35,12 +63,26 @@ function technic.register_generator(data)
 	for k, v in pairs(groups) do active_groups[k] = v end
 
 	local generator_formspec =
-		"size[8,9;]"..
+		size..
 		"label[0,0;"..S("Fuel-Fired @1 Generator", S(tier)).."]"..
 		"list[context;src;3,1;1,1;]"..
-		"image[4,1;1,1;default_furnace_fire_bg.png]"..
-		"list[current_player;main;0,5;8,4;]"..
-		"listring[]"
+		"listring[context;src]"..
+		"image[4,1;1,1;default_furnace_fire_bg.png]"
+
+	if minetest.get_modpath("mcl_formspec") then
+		generator_formspec = generator_formspec..
+			mcl_formspec.get_itemslot_bg(3,1,1,1)..
+			-- player inventory
+			"list[current_player;main;0,5.5;9,3;9]"..
+			mcl_formspec.get_itemslot_bg(0,5.5,9,3)..
+			"list[current_player;main;0,8.74;9,1;]"..
+			mcl_formspec.get_itemslot_bg(0,8.74,9,1)..
+			"listring[current_player;main]"
+	else
+		generator_formspec = generator_formspec..
+			"list[current_player;main;0,5;8,4;]"..
+			"listring[current_player;main]"
+	end
```

Comment:
Delete lines 65-85 and utilize `update_generator_formspec`:
This isn't really needed anymore now that you've added `update_generator_formspec`. Better just delete this and use `update_generator_formspec` function for `on_construct` and `on_receive_fields`.

Makes changes a lot cleaner (no dupe formspec defs) and those are least expensive events anyway so having small part cached isn't really useful.

Also if you want I can help a bit with these simple cleanups.

Reply:
I don't understand what you mean
that function could not be used as `on_construct` or `on_receive_fields`

Reply:
> I don't understand what you mean
> that function could not be used as `on_construct` or `on_receive_fields`

Just simply call it like everywhere else and remove formspec dupes, add condition for image filter if needed (not sure if it is needed, probably not).

Reply:
changed